File: backend/src/ws.rs
use std::sync::{Arc, Mutex};

use actix_web::{web, HttpRequest, HttpResponse};
use actix_ws::{Message, MessageStream, Session};
use futures_util::StreamExt;
use serde::Deserialize;
use uuid::Uuid;

use crate::views;
use crate::AppState;

const FRAME_INTERVAL_MS: u64 = 1000;
const EXTENDED_EVERY: u64 = 3;

/// Per-connection subscription preferences, mutated by the receiver task and
/// read by the sender task.
#[derive(Default)]
struct Prefs {
    symbol: Mutex<String>,
    agent_id: Mutex<Option<Uuid>>,
}

impl Prefs {
    fn new(symbol: Option<String>, agent_id: Option<Uuid>) -> Self {
        Prefs {
            symbol: Mutex::new(symbol.unwrap_or_else(|| "NOVA".into())),
            agent_id: Mutex::new(agent_id),
        }
    }
}

#[derive(Deserialize)]
pub struct WsQuery {
    symbol: Option<String>,
    agent_id: Option<String>,
}

pub async fn index(
    req: HttpRequest,
    body: web::Payload,
    state: web::Data<AppState>,
    q: web::Query<WsQuery>,
) -> HttpResponse {
    let agent_id = q.agent_id.as_deref().and_then(|s| Uuid::parse_str(s).ok());
    let (response, session, stream) = match actix_ws::handle(&req, body) {
        Ok(v) => v,
        Err(_) => return HttpResponse::BadRequest().finish(),
    };
    let prefs = Arc::new(Prefs::new(q.symbol.clone(), agent_id));

    actix_web::rt::spawn(sender_task(session.clone(), state.clone(), prefs.clone()));
    actix_web::rt::spawn(receiver_task(session, stream, prefs));
    response
}

async fn sender_task(mut session: Session, state: web::Data<AppState>, prefs: Arc<Prefs>) {
    use std::time::Duration;
    let mut interval = actix_web::rt::time::interval(Duration::from_millis(FRAME_INTERVAL_MS));
    let mut seq: u64 = 0;

    loop {
        interval.tick().await;
        let symbol = prefs.symbol.lock().unwrap_or_else(|p| p.into_inner()).clone();
        let agent_id = *prefs.agent_id.lock().unwrap_or_else(|p| p.into_inner());
        seq += 1;
        let extended = seq % EXTENDED_EVERY == 0;

        let frame = {
            let ex = state.lock();
            views::build_frame(&ex, &symbol, agent_id, extended, seq)
        };
        match serde_json::to_string(&frame) {
            Ok(json) => {
                if session.text(json).await.is_err() {
                    break; // peer went away
                }
            }
            Err(e) => log::error!("ws frame encode failed: {e}"),
        }
    }
}

#[derive(Deserialize)]
struct ClientMsg {
    #[serde(rename = "type")]
    kind: String,
    symbol: Option<String>,
    agent_id: Option<String>,
}

async fn receiver_task(mut session: Session, mut stream: MessageStream, prefs: Arc<Prefs>) {
    while let Some(Ok(msg)) = stream.next().await {
        match msg {
            Message::Ping(bytes) => {
                if session.pong(&bytes).await.is_err() {
                    break;
                }
            }
            Message::Text(text) => match serde_json::from_str::<ClientMsg>(&text) {
                Ok(m) if m.kind == "subscribe" => {
                    if let Some(s) = m.symbol {
                        *prefs.symbol.lock().unwrap_or_else(|p| p.into_inner()) = s;
                    }
                    let aid = m
                        .agent_id
                        .as_deref()
                        .and_then(|s| Uuid::parse_str(s).ok());
                    *prefs.agent_id.lock().unwrap_or_else(|p| p.into_inner()) = aid;
                    let ack = serde_json::json!({
                        "type": "subscribed",
                        "symbol": prefs.symbol.lock().map(|g| g.clone()).ok(),
                        "agent_id": aid,
                    });
                    let _ = session.text(ack.to_string()).await;
                }
                Ok(m) if m.kind == "ping" => {
                    let _ = session.text(r#"{"type":"pong"}"#.to_string()).await;
                }
                _ => { /* unknown or malformed messages are ignored */ }
            },
            Message::Close(_) => break,
            _ => {}
        }
    }
    let _ = session.close(None).await;
}

File: backend/ws.go
package main

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"sync"
	"time"

	"github.com/coder/websocket"
	"github.com/google/uuid"
)

const (
	frameIntervalMs = 1000
	extendedEvery   = 3
)

// prefs holds per-connection subscription preferences, mutated by the
// receiver goroutine and read by the sender goroutine.
type prefs struct {
	mu      sync.Mutex
	symbol  string
	agentID *uuid.UUID
}

func newPrefs(symbol string, agentID *uuid.UUID) *prefs {
	if symbol == "" {
		symbol = "NOVA"
	}
	return &prefs{symbol: symbol, agentID: agentID}
}

// wsClientMsg mirrors the client->server protocol messages.
type wsClientMsg struct {
	Type    string  `json:"type"`
	Symbol  *string `json:"symbol"`
	AgentID *string `json:"agent_id"`
}

func (s *server) handleWS(w http.ResponseWriter, r *http.Request) {
	c, err := websocket.Accept(w, r, nil)
	if err != nil {
		return
	}

	agentID := (*uuid.UUID)(nil)
	if v := r.URL.Query().Get("agent_id"); v != "" {
		if id, err := uuid.Parse(v); err == nil {
			agentID = &id
		}
	}
	p := newPrefs(r.URL.Query().Get("symbol"), agentID)

	ctx, cancel := context.WithCancel(r.Context())
	writeMu := &sync.Mutex{}
	write := func(msgType websocket.MessageType, payload []byte) {
		writeMu.Lock()
		defer writeMu.Unlock()
		ctx, cxl := context.WithTimeout(ctx, 5*time.Second)
		defer cxl()
		if err := c.Write(ctx, msgType, payload); err != nil {
			cancel()
		}
	}

	go s.wsSender(ctx, c, p, write, cancel)
	s.wsReceiver(ctx, c, p, write, cancel)
	_ = c.Close(websocket.StatusNormalClosure, "")
}

// wsSender pushes a full snapshot frame every second.
func (s *server) wsSender(ctx context.Context, c *websocket.Conn, p *prefs, write func(websocket.MessageType, []byte), cancel context.CancelFunc) {
	ticker := time.NewTicker(frameIntervalMs * time.Millisecond)
	defer ticker.Stop()
	var seq uint64
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
		p.mu.Lock()
		symbol := p.symbol
		agentID := p.agentID
		p.mu.Unlock()

		seq++
		extended := seq%extendedEvery == 0

		ex := s.ex.lock()
		frame := BuildFrame(ex, symbol, agentID, extended, seq)
		s.ex.unlock()

		payload, err := json.Marshal(frame)
		if err != nil {
			slog.Error("ws frame encode failed", "err", err)
			continue
		}
		write(websocket.MessageText, payload)
	}
}

// wsReceiver handles client messages: subscribe updates prefs + acks, ping
// gets a pong. Protocol-level ping/pong control frames are handled inside
// coder/websocket automatically.
func (s *server) wsReceiver(ctx context.Context, c *websocket.Conn, p *prefs, write func(websocket.MessageType, []byte), cancel context.CancelFunc) {
	defer cancel()
	for {
		_, data, err := c.Read(ctx)
		if err != nil {
			return
		}
		var msg wsClientMsg
		if err := json.Unmarshal(data, &msg); err != nil {
			continue // unknown or malformed messages are ignored
		}
		switch msg.Type {
		case "subscribe":
			p.mu.Lock()
			if msg.Symbol != nil {
				p.symbol = *msg.Symbol
			}
			agentID := p.agentID
			if msg.AgentID != nil {
				if id, err := uuid.Parse(*msg.AgentID); err == nil {
					agentID = &id
				} else {
					agentID = nil
				}
				p.agentID = agentID
			}
			ackSymbol := p.symbol
			p.mu.Unlock()
			ack, _ := json.Marshal(map[string]any{
				"type":     "subscribed",
				"symbol":   ackSymbol,
				"agent_id": agentID,
			})
			write(websocket.MessageText, ack)
		case "ping":
			write(websocket.MessageText, []byte(`{"type":"pong"}`))
		}
	}
}
